### PR TITLE
Support externally-provisioned runs in cloud run --local-execution

### PR DESCRIPTION
**Pull request by ankur22**
> **Supersedes #6133.** Now based directly on `master` following the merge of the provisioning migration (#6169, which superseded #6068); adapted to the int64 resource-ID SDK change.

## What?

Follow-up to the provisioning migration (#6169, now merged). Let `k6 cloud run --local-execution` push metrics to a run that an **external service** has already provisioned via the provisioning API, instead of provisioning the run itself.

When an external orchestrator calls `start_local_execution`, then launches local k6 with `K6_CLOUD_PUSH_REF_ID` set (the run id), k6 must push metrics to that run's scoped `push_url` with the scoped `test_run_token`. Today that combination silently falls back to the legacy ingest endpoint with the long-lived org token. Two changes fix it:

1. `cloudapi.Config.MetricsPushURL` and `TestRunToken` gain `envconfig` tags (`K6_CLOUD_METRICS_PUSH_URL`, `K6_CLOUD_TEST_RUN_TOKEN`) so an orchestrator can supply the scoped push credentials via env. They were previously JSON-only, set solely by k6's own provisioning call.
2. The cloud Output's provisioning-mode gate now arms whenever `MetricsPushURL` + `TestRunToken` are present, even when `PushRefID` is set (previously `PushRefID` disabled it). In that externally-provisioned case k6 pushes to the provisioning URL with the scoped token but does **not** construct a notifier — the external service owns the run lifecycle (create/start/notify), which `testFinished` already respects by short-circuiting on `PushRefID`. A half-set credential pair (one env var without the other) is rejected rather than silently using the legacy path.

Behaviour matrix:

| Signal | Push endpoint | Token | Notify |
|---|---|---|---|
| `PushRefID` only (unchanged) | legacy ingest | long-lived | no |
| `PushRefID` + scoped creds (new) | provisioning `push_url` | scoped `test_run_token` | no |
| self-provision, no ref (unchanged) | provisioning `push_url` | scoped | yes |

`k6 run --out cloud` (path B), self-provisioned `--local-execution`, and the plain PushRefID flow are all **unchanged**. Cloud secrets need no code change — an orchestrator configures them via the existing URL secret source env vars (`--secret-source=url` + `K6_SECRET_SOURCE_URL_*`), or passes `--no-cloud-secrets`.

Two atomic commits, each individually buildable + lintable + green.

## Why?

Orchestration services (and future flows such as Synthetic Monitoring) provision the run themselves via the provisioning API and hand local k6 a correlation ref rather than letting k6 self-provision. The provisioning API already returns a per-run `push_url` and a scoped `test_run_token` for exactly this; without this change that externally-provisioned combination falls back to the legacy v1 ingest endpoint with the long-lived org token — defeating the scoped-token model and forcing the long-lived token to be distributed to every local runner.

## Checklist

- [x] I have performed a self-review of my code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests for my changes.
- [x] I have run linter and tests locally (`make check`) and all pass.

## Related PR(s)/Issue(s)

- Follows #6169 (the provisioning migration — merged into `master`; supersedes #6068 and #6133). This PR is now based directly on `master` and contains only the two external-provisioning commits.
- Verified end-to-end against a real stack with a dedicated `external-provisioned` manual-test scenario (out-of-band `start_local_execution`, then assert metrics push to `push_url` with `Authorization: Bearer <test_run_token>`, and no self-provision / poll / notify / legacy `/v1/tests`).

### CODE REVIEW

File: cloudapi/config_test.go [hunk 1]
```diff
@@ -57,13 +57,12 @@ func TestConfigApply(t *testing.T) {
...
-// and NOT on lib.Options (which is serialised into the archive's
-// metadata.json and pushed to the cloud). A new field accidentally added
-// to lib.Options would be serialised into the archive — but the new
-// fields here are on cloudapi.Config, so they cannot leak that way.
+// The scoped test_run_token is short-lived and run-scoped; it is now
+// deliberately accepted via env (K6_CLOUD_TEST_RUN_TOKEN) so an
+// external orchestrator that provisioned the run can hand it to a
+// local k6. It still never touches lib.Options / the archive — the
+// long-lived org token remains the sensitive credential and is
+// unaffected here.
```
Comment (joanlopez) [inline cloudapi/config_test.go]: Is this change really intended? I see the code comment no longer starts with the test name, which I think is a common convention, but not only that, but I see no changes in the test, which confuses me even more 😅 🤔

File: cloudapi/config_test.go [hunk 2]
```diff
@@ -85,22 +84,23 @@ func TestConfig_NewFieldsRoundTripThroughJSON(t *testing.T) {
 	assert.Equal(t, original.TestRunToken, roundTripped.TestRunToken)
 }
 
-func TestConfig_NewFieldsNotPickedUpByEnvconfig(t *testing.T) {
+// TestConfig_NewFieldsFromEnvconfig verifies the scoped push creds are
+// read from the environment via envconfig (K6_CLOUD_METRICS_PUSH_URL /
+// K6_CLOUD_TEST_RUN_TOKEN) so an external orchestrator that provisioned
+// a run can hand them to a local k6.
+func TestConfig_NewFieldsFromEnvconfig(t *testing.T) {
```
Comment (joanlopez) [inline cloudapi/config_test.go]: Should we somehow test both scenarios separately? 
- Self provisioned
- Externally provisioned

I'm wondering whether an existing `K6_CLOUD_TEST_RUN_TOKEN` variable might still override the value set at runtime in the self-provisioned path, and break it, which I guess could not only be considered technically a breaking change, but perhaps also mislead users due to a poor UX. Hereby, we might want to write a warning log in such case, I guess. If so, I guess it might make sense to also add a test for that.

**Comment by joanlopez** [comment]
Out of curiosity,

> Today that combination silently falls back to the legacy ingest endpoint with the long-lived org token.

Does this mean that this is technically considered a bug fix? I'm wondering whether we should take into consideration if this may "break" some behavior to existing users.

File: cloudapi/config_test.go (hunk 2 above)
Comment (joanlopez) [reply cloudapi/config_test.go]: I can confirm `K6_CLOUD_TEST_RUN_TOKEN=123 k6 cloud run <script> --local-execution` now fails, while it used to work, which I'm not sure if that's the desired outcome.

Comment (joanlopez) [reply cloudapi/config_test.go]: I'm also concerned about other scenarios that are possible now, for instance, stale credentials in the k6 config file. Perhaps we should be more defensive against all these corner cases? 🤔

**Comment by joanlopez** [comment]
> The provisioning API already returns a per-run push_url and a scoped test_run_token for exactly this; without this change that externally-provisioned combination falls back to the legacy v1 ingest endpoint with the long-lived org token — defeating the scoped-token model and forcing the long-lived token to be distributed to every local runner.

Does this mean that, with the changes in this pull; stack & token should no longer be required to run? Because I think that, as-is now, even by providing `K6_CLOUD_METRICS_PUSH_URL`, `K6_CLOUD_TEST_RUN_TOKEN`, and `K6_CLOUD_PUSH_REF_ID` it still requires auth settings.

OTOH, I guess there's still features (e.g., tracing? `insights.NewDefaultClientConfigForTestRun(TracesHost, 
  o.config.Token.String, testRunID)`) that would need to be migrated before not requiring auth settings at all.

**Review by joanlopez — COMMENTED** [review]
I think the hot path coverage looks good, but I'm still dubious whether the coverage for corner cases is good enough to get this merged. I left some comments related.

File: cloudapi/config_test.go (hunk 2 above)
Comment (ankur22) [reply cloudapi/config_test.go]: What do you think of these set of e2e tests:

|#|	Command (run after the common exports)|
| --- | --- |
|A1|	k6 cloud run --local-execution test.js|
|A2|	K6_CLOUD_TEST_RUN_TOKEN=bogus-env-token k6 cloud run --local-execution test.js|
|A3|	K6_CLOUD_METRICS_PUSH_URL=$H/bogus-metrics k6 cloud run --local-execution test.js|
|A4|	K6_CLOUD_TEST_RUN_TOKEN=bogus-env-token K6_CLOUD_METRICS_PUSH_URL=$H/bogus-metrics k6 cloud run --local-execution test.js|
|A5|	K6_CLOUD_TEST_RUN_TOKEN=bogus-env-token k6 run --out=cloud test.js|
|A6|	K6_CLOUD_PUSH_REF_ID=99999 k6 cloud run --local-execution test.js|
|A7|	K6_CLOUD_PUSH_REF_ID=1337 K6_CLOUD_TEST_RUN_TOKEN=ext-scoped-token k6 run --out=cloud test.js|
|A8|	k6 cloud run --local-execution test.js (+ config file below)|
|B1|	K6_CLOUD_PUSH_REF_ID=99999 K6_CLOUD_METRICS_PUSH_URL=$H/v1/metrics K6_CLOUD_TEST_RUN_TOKEN=ext-scoped-token k6 cloud run --local-execution test.js|
|B2|	K6_CLOUD_PUSH_REF_ID=99999 K6_CLOUD_TEST_RUN_TOKEN=ext-scoped-token k6 cloud run --local-execution test.js|

```json
{"collectors":{"cloud":{"testRunToken":"stale-config-token"}}}
```

File: cloudapi/config_test.go (hunk 1 above)
Comment (ankur22) [reply cloudapi/config_test.go]: Good catch — that was unintended churn, sorry. The comment only changed because this PR had (temporarily) made those fields env-settable, which broke the original *"scoped tokens stay in-process, never env-settable"* framing, so I'd rewritten it.

Since we've now dropped the env-binding approach entirely (see the thread below / the `Fix env override…` commit), the fields are programmatic-only again and that original note is accurate once more. I've reverted both the comment and the companion `TestConfig_NewFieldsNotPickedUpByEnvconfig` back to their original form, so this diff no longer shows up.

File: cloudapi/config_test.go (hunk 2 above)
Comment (ankur22) [reply cloudapi/config_test.go]: You were right on every count — this was a real regression, now fixed in 0c6c4bb9.

**Root cause:** binding `MetricsPushURL`/`TestRunToken` to env meant the cloud output re-consolidates its config with env applied *last*, so a stray `K6_CLOUD_TEST_RUN_TOKEN` (or a stale value in the k6 config file) overrode the run-scoped values from `start_local_execution` — exactly the `K6_CLOUD_TEST_RUN_TOKEN=123 …` failure you reproduced.

**Fix — remove the footgun rather than warn:** dropped the two `envconfig` tags (fields are programmatic-only again). The externally-provisioned path now reads `K6_CLOUD_METRICS_PUSH_URL`/`K6_CLOUD_TEST_RUN_TOKEN` explicitly in the `PushRefID` branch (`applyExternalProvisioningCreds`), requiring both-or-neither. The self-provisioned path never consults those env vars anymore, so there's no override left to warn about — and a stale config-file value can't hijack it either.

I implemented the matrix above as `TestExtProvMatrix` and ran it on `master`, the previous (broken) commit, and the fix:

| # | master | broken | fix |
|---|:--:|:--:|:--:|
| A1 | ✓ | ✓ | ✓ |
| A2 | ✓ | ✗ | ✓ |
| A3 | ✓ | ✗ | ✓ |
| A4 | ✓ | ✗ | ✓ |
| A5 | ✓ | ✓ | ✓ |
| A6 | ✓ | ✓ | ✓ |
| A7 | ✓ | ✗ | ✓ |
| A8 | ✓ | ✓ | ✓ |
| B1 | ✗* | ✓ | ✓ |
| B2 | ✗* | ✓ | ✓ |

So the self-provisioned/legacy regressions (A2/A3/A4/A7) flip red→green, the externally-provisioned feature (B1) is preserved, partial creds now give a clear `must be set together` error (B2), and the stale-config-file case (A8) is covered. (*B1/B2 "fail" on `master` only because the feature doesn't exist there yet — expected.)

**Comment by ankur22** [comment]
> Does this mean that this is technically considered a bug fix? I'm wondering whether we should take into consideration if this may "break" some behavior to existing users.

I'd call it a new capability rather than a bug fix: the old behavior (externally-provisioned run falling back to the legacy ingest endpoint with the org token) *worked* — it just didn't use the per-run scoped-token model. This PR lets that combination use the scoped push URL + token instead.

On breaking existing users: the new path only triggers when the **new** env vars `K6_CLOUD_METRICS_PUSH_URL` + `K6_CLOUD_TEST_RUN_TOKEN` are set *together with* `K6_CLOUD_PUSH_REF_ID`. Those two vars did nothing before this PR, so nobody is relying on that combination today. Existing `PushRefID`-only users keep the exact legacy org-token push (scenario A6 guards this). So I don't think there's a behavior change for anyone currently.

The one genuine regression — a stray/stale `K6_CLOUD_TEST_RUN_TOKEN` breaking the *self-provisioned* path — is the bug you caught in the `config_test.go` thread, now fixed and guarded.

**Comment by ankur22** [comment]
> Does this mean that, with the changes in this pull; stack & token should no longer be required to run? ... even by providing `K6_CLOUD_METRICS_PUSH_URL`, `K6_CLOUD_TEST_RUN_TOKEN`, and `K6_CLOUD_PUSH_REF_ID` it still requires auth settings.

Correct, and it's intentional here: `checkCloudLogin` still runs before the `PushRefID` branch, so `K6_CLOUD_TOKEN` + `K6_CLOUD_STACK_ID` remain required even in the externally-provisioned case. The scoped creds only govern where/how metrics (and logs, in the stacked PR) are pushed — they're not meant to replace login.

And you've pinpointed exactly why we can't drop the auth requirement yet: other paths still use the long-lived org token — insights/tracing (`insights.NewDefaultClientConfigForTestRun(TracesHost, o.config.Token.String, …)` in `output/cloud/expv2/output.go`) and the secrets source. Removing the stack+token requirement would mean migrating those off the org token too, which is out of scope for this PR. So this one deliberately keeps login intact and just closes the metrics-push gap; making auth optional would be a good follow-up once tracing/secrets are migrated.

File: internal/cmd/tests/cmd_cloud_run_extprov_matrix_test.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+	// (e.g. a clobbered URL landing on the catch-all).
+	srv.Mux.HandleFunc("/v1/metrics", rec.handler)
+	srv.Mux.HandleFunc("/", rec.handler)
+
+	ts.Env["K6_CLOUD_HOST"] = srv.URL
+	ts.Env["K6_CLOUD_HOST_V6"] = srv.URL
+	return ts, srv, rec, &notified
+}
+
+func TestExtProvMatrix(t *testing.T) {
```
Comment (ankur22) [inline internal/cmd/tests/cmd_cloud_run_extprov_matrix_test.go]: I wasn't going to commit this, but actually it feels useful and protects against regressions. Also documents all edge cases for this particular feature (--local-execution), although it isn't exhaustive (not taking into account other flags etc.)

File: internal/cmd/tests/cmd_cloud_run_push_credentials_test.go [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+//     and token via K6_CLOUD_METRICS_PUSH_URL and K6_CLOUD_TEST_RUN_TOKEN,
+//     which are required together.
+//   - Legacy (k6 run --out cloud, or a PushRefID relay without scoped
+//     creds): metrics go to the endpoint derived from the host using the
+//     long-lived org token.
+//
+// Each case runs the real command against an in-process mock cloud
+// (httptest, no network) and asserts which bearer token k6 pushes with,
+// using distinct sentinel tokens so the source is unambiguous.
+func TestCloudMetricsPushCredentials(t *testing.T) {
```
Comment (ankur22) [inline internal/cmd/tests/cmd_cloud_run_push_credentials_test.go:184]: I wasn't going to commit this, but actually it feels useful and protects against regressions. Also documents all edge cases for this particular feature (--local-execution), although it isn't exhaustive (not taking into account other flags etc.)

Comment (janHildebrandt98) [reply internal/cmd/tests/cmd_cloud_run_push_credentials_test.go:184]: I really like these tests. Nice that you kept them.

**Comment by janHildebrandt98** [comment]
LGTM 🫡

**Review by janHildebrandt98 — APPROVED** [review]

File: internal/cmd/tests/cmd_cloud_run_push_credentials_test.go (hunk 4 above)
Comment (joanlopez) [reply internal/cmd/tests/cmd_cloud_run_push_credentials_test.go:184]: Indeed, I like them! 👍🏻

File: internal/output/cloud/output.go [hunk 5]
```diff
@@ -223,11 +223,22 @@ func (out *Output) Start() error {
...
-		// future cmd-layer regression that populates both fields.
-		if out.config.MetricsPushURL.Valid && out.config.TestRunToken.Valid && !out.config.PushRefID.Valid {
+		// Exactly one of the scoped push creds set is a misconfiguration.
+		// The cmd layer already validates this for the externally-provisioned
+		// env case (internal/cmd/outputs_cloud.go); this is the defensive
+		// backstop for any other source (e.g. a hand-written cloud config).
+		if out.config.MetricsPushURL.Valid != out.config.TestRunToken.Valid {
+			return fmt.Errorf(
+				"both K6_CLOUD_METRICS_PUSH_URL and K6_CLOUD_TEST_RUN_TOKEN " +
+					"must be set together")
```
Comment (joanlopez) [inline internal/output/cloud/output.go:233]: nit; I think this should be `errors.New(...)` as there's no argument nor anything to format.

**Review by joanlopez — APPROVED** [review]
Great job with the new test suite! 🚀

**Review by szkiba — APPROVED** [review]
